doc/HLD_of_FOL.rst

```diff
@@ -0,0 +1,290 @@
+
+Record Structure
+=================
+
+A fol record, added via the main fol interface, contains the following:
+
+1[R.BACK-END.TRANSACTIONAL] ST
+
+2[R.BACK-END.INDEXING] ST
+
+- an operation opcode, identifying the type of file system operation;
+
+- lsn;
+
+- information sufficient to undo and redo the update, described by the record, including:
+
+  - for each file system object affected by the update, its identity (a fid) and its object version identifying the state of the object in which the update can be applied;
+
+  - any additional operation type dependent information (file names, attributes, etc.) necessary to execute or roll-back the update;
+
+- information sufficient to identify other updates of the same operation (if any) and their state. For the purposes of the present design specification it's enough to posit that this can be done by means of some opaque identifier;
+
+- for each object modified by the update, a reference (in the form of lsn) to the record of the previous update to this object (null is the update is object creation). This reference is called prev-lsn reference;
+
+- distributed transaction management data, including an epoch this update and operation are parts of;
+
+- liveness state: a number of outstanding references to this record
+
+Liveness and Pruning
+=====================
+
+A node fol must be prunable if only to function correctly on a node without persistent storage. At the same time, a variety of sub-systems both from M0 core and outside of it, might want to refer to fol records. To make pruning possible and flexible, each fol record is augmented with a reference counter, counting all outstanding references to the record. A record can be pruned iff its reference counter drops to 0 together with reference counters of all earlier (in lsn sense) unpruned records in the fol.
+
+Conformance
+=============
+
+- [r.fol.every-node]: on nodes with persistent storage, M0 core runs in the user space and the fol is stored in a data-base table. On a node without persistent storage, M0 core runs in the kernel space and the fol is stored in memory-only index. Data-base and memory-only index provide the same external interface, making fol code portable;
+
+- [r.fol.local-txn]: request handler inserts a record into FOL table in the context of the same transaction where update is executed. This guarantees WAL property of fol;
+
+- [R.FOL]: vacuous;
```

Comment:
The title in the [] should use [CAPITAL.LETTER], just like those in other places.

All the word "fol" must be converted to upper case "FOL".

Reply:
Incorporated.